File: AgoraChat/Code/Controllers/Settings/NotificationsSetting/ACDNotificationSettingViewController.m
//
//  ACDNotificationSettingViewController.m
//  AgoraChat
//
//  Created by hxq on 2022/3/16.
//  Copyright © 2022 easemob. All rights reserved.
//

#import "ACDNotificationSettingViewController.h"
#import "ACDSilentModeSetViewController.h"
#import "ACDTitleDetailCell.h"
#import "ACDSubDetailCell.h"
#import "ACDNameSwitchCell.h"
#import "ACDDateHelper.h"

@interface ACDNotificationSettingViewController ()
@property (nonatomic , strong) AgoraChatSilentModeResult *silentModeItem;
@property (nonatomic , strong) NSArray *cells;
@property (nonatomic , strong) ACDTitleDetailCell *remindTypeCell;
@property (nonatomic , strong) ACDSubDetailCell *muteCell;
@property (nonatomic , strong) ACDNameSwitchCell *showPreTextCell;
@property (nonatomic , strong) ACDNameSwitchCell *soundCell;
@property (nonatomic , strong) ACDNameSwitchCell *vibrateCell;
@property (nonatomic , copy) NSString *navTitle;
@property (nonatomic , copy) NSString *muteCellNameTitle;
@property (nonatomic , copy) NSString *remindCellNameTitle;
@end

@implementation ACDNotificationSettingViewController

- (void)viewDidLoad {
    [super viewDidLoad];
   
    self.automaticallyAdjustsScrollViewInsets = NO;
    if(@available(iOS 15.0,*) ){
        self.table.sectionHeaderTopPadding = 0.0f;
    }
    [self setNavBar];
    [self getDataFromSever];
    
}

- (void)setNavBar {
    UIButton *backButton = [[UIButton alloc] initWithFrame:CGRectMake(0, 0, 8, 15)];
    [backButton setImage:[UIImage imageNamed:@"black_goBack"] forState:UIControlStateNormal];
    [backButton addTarget:self action:@selector(backAction) forControlEvents:UIControlEventTouchUpInside];
    
    [backButton setTitle:self.navTitle forState:UIControlStateNormal];
    [backButton setTitleColor:TextLabelBlackColor forState:UIControlStateNormal];
    
    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithCustomView:backButton];
}

- (void)backAction {
    [self.navigationController popViewControllerAnimated:YES];
}
- (void)getDataFromSever
{
    [self hideHud];
    [self showHudInView:self.view hint:NSLocalizedString(@"hud.load", @"Loading..")];
    ACD_WS
    if (self.notificationType == AgoraNotificationSettingTypeSelf) {
        [[AgoraChatClient sharedClient].pushManager getSilentModeForAllWithCompletion:^(AgoraChatSilentModeResult * _Nonnull aResult, AgoraChatError * _Nonnull aError) {
            [weakSelf hideHud];
            if (!aError) {
                weakSelf.silentModeItem = aResult;
                [weakSelf.table reloadData];
            }else{
                [weakSelf showHint:aError.errorDescription];
            }
        }];
    }else{
        
        AgoraChatConversationType chatType = AgoraChatConversationTypeGroupChat;
        if (self.notificationType == AgoraNotificationSettingTypeSingleChat) {
            chatType = AgoraChatConversationTypeChat;
        }
        [[AgoraChatClient sharedClient].pushManager getSilentModeForConversation:self.conversationID conversationType:chatType completion:^(AgoraChatSilentModeResult * _Nonnull aResult, AgoraChatError * _Nonnull aError) {
            [weakSelf hideHud];
            if (!aError) {
                weakSelf.silentModeItem = aResult;
                [weakSelf.table reloadData];
            }else{
                [weakSelf showHint:NSLocalizedString(@"hud.fail", @"fail")];
            }
        }];
    }
}

#pragma mark - action
- (void)remindTypeAction {
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:nil message:nil preferredStyle:UIAlertControllerStyleActionSheet];
    
    if (self.notificationType != AgoraNotificationSettingTypeSelf) {
        UIAlertAction *defaultAction = [UIAlertAction actionWithTitle:@"Default" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            [self remindTypeDefault];
        }];
        [alertController addAction:defaultAction];
        [defaultAction setValue:TextLabelBlackColor forKey:@"titleTextColor"];
    }
    
    UIAlertAction *allAction = [UIAlertAction actionWithTitle:@"All Messages" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self remindTypeChange:AgoraChatPushRemindTypeAll];
    }];
    [alertController addAction:allAction];
    [allAction setValue:TextLabelBlackColor forKey:@"titleTextColor"];
    
    if(self.notificationType != AgoraNotificationSettingTypeSingleChat)
    {
        UIAlertAction *metionAction = [UIAlertAction actionWithTitle:@"Only @Metions" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            [self remindTypeChange:AgoraChatPushRemindTypeMentionOnly];
        }];
        [alertController addAction:metionAction];
        [metionAction setValue:TextLabelBlackColor forKey:@"titleTextColor"];
    }
    UIAlertAction *noneAction = [UIAlertAction actionWithTitle:@"Nothing" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self remindTypeChange:AgoraChatPushRemindTypeNone];
    }];
    [alertController addAction:noneAction];
    
    [noneAction setValue:TextLabelBlackColor forKey:@"titleTextColor"];
    
    [alertController addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
    }]];
    
    [self presentViewController:alertController animated:YES completion:nil];
    
}

- (void)muteAction {
    NSTimeInterval timestampNow = [[NSDate date] timeIntervalSince1970] * 1000;
    if (self.silentModeItem.expireTimestamp > timestampNow) {
        //先弹窗关闭再进去设置
        NSString *titleStr = @"Turn off Do Not Disturb?";
        NSString *infoStr = [NSString stringWithFormat:@"You have set Do Not Disturb until \n %@",self.muteCell.subDetailLabel.text];
        
        if (self.notificationType != AgoraNotificationSettingTypeSelf) {
            titleStr = [NSString stringWithFormat:@"Unmute this %@?",self.notificationType == AgoraNotificationSettingTypeGroup?@"Group":@"Thread"];
            infoStr = [NSString stringWithFormat:@"You have muted this %@ until \n %@",self.notificationType == AgoraNotificationSettingTypeGroup?@"Group":@"Thread",self.muteCell.subDetailLabel.text];
        }
        UIAlertController *alertController = [UIAlertController alertControllerWithTitle:titleStr message:infoStr preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
            
        }];
        [alertController addAction:cancelAction];
        
        UIAlertAction *sureAction = [UIAlertAction actionWithTitle:@"OK" style: UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            [self closeMute];
        }];
        // [deleteAction setValue:TextLabelPinkColor forKey:@"titleTextColor"];
        
        [alertController addAction:sureAction];
        
        alertController.modalPresentationStyle = 0;
        [self presentViewController:alertController animated:YES completion:nil];
    }else {
        ACDSilentModeSetViewController *controller = [[ACDSilentModeSetViewController alloc]init];
        controller.conversationID = self.conversationID;
        controller.notificationType = self.notificationType;
        ACD_WS
        controller.doneBlock = ^(AgoraChatSilentModeResult * _Nonnull item) {
            weakSelf.silentModeItem = item;
            [weakSelf.table reloadData];
        };
        [self.navigationController pushViewController:controller animated:YES];
    }
    
}


- (void)closeMute
{
    AgoraChatSilentModeParam *param = [[AgoraChatSilentModeParam alloc] initWithParamType:AgoraChatSilentModeParamTypeDuration];
    param.silentModeDuration = 0;
    [self silentModeChange:param];
    
}

- (void)remindTypeChange:(AgoraChatPushRemindType)remindType {
    
    AgoraChatSilentModeParam *param = [[AgoraChatSilentModeParam alloc] initWithParamType:AgoraChatSilentModeParamTypeRemindType];
    param.remindType = remindType;
    [self silentModeChange:param];
  
}
- (void)silentModeChange:(AgoraChatSilentModeParam*)param
{
    ACD_WS
    if (self.notificationType == AgoraNotificationSettingTypeSelf) {
        [[AgoraChatClient sharedClient].pushManager setSilentModeForAll:param completion:^(AgoraChatSilentModeResult * _Nonnull aResult, AgoraChatError * _Nonnull aError) {
            if (!aError) {
                weakSelf.silentModeItem = aResult;
                [weakSelf.table reloadData];
            }else{
                [weakSelf showHint:NSLocalizedString(@"hud.fail", @"Fail")];
            }
        }];
    }else{
        AgoraChatConversationType type = AgoraChatConversationTypeGroupChat;
        if (self.notificationType == AgoraNotificationSettingTypeSingleChat) {
            type = AgoraChatConversationTypeChat;
        }
        [[AgoraChatClient sharedClient].pushManager setSilentModeForConversation:self.conversationID conversationType:type params:param completion:^(AgoraChatSilentModeResult * _Nonnull aResult, AgoraChatError * _Nonnull aError) {
            if (!aError) {
                weakSelf.silentModeItem = aResult;
                [weakSelf.table reloadData];
            }else{
                [weakSelf showHint:NSLocalizedString(@"hud.fail", @"Fail")];
            }
        }];
    }
}
- (void)remindTypeDefault {
    [self hideHud];
    [self showHudInView:self.view hint:NSLocalizedString(@"hud.wait", @"Waiting..")];
    ACD_WS
    AgoraChatConversationType type = AgoraChatConversationTypeGroupChat;
    if (self.notificationType == AgoraNotificationSettingTypeSingleChat) {
        type = AgoraChatConversationTypeChat;
    }
    [[AgoraChatClient sharedClient].pushManager clearRemindTypeForConversation:self.conversationID conversationType:type completion:^(AgoraChatSilentModeResult * _Nonnull aResult, AgoraChatError * _Nonnull aError) {
        [weakSelf hideHud];
        if (!aError) {
            weakSelf.silentModeItem = aResult;
            [weakSelf.table reloadData];
            
        }else{
            [weakSelf showHint:aError.errorDescription];
        }
    }];
}


- (void)showPreTextAction {
    AgoraChatPushDisplayStyle style = AgoraChatPushDisplayStyleSimpleBanner;
    if (self.showPreTextCell.aSwitch.on) {
        style = AgoraChatPushDisplayStyleMessageSummary;
    }
    ACD_WS
    [[AgoraChatClient sharedClient].pushManager updatePushDisplayStyle:style completion:^(AgoraChatError * _Nonnull aError) {
        if (aError) {
            [weakSelf showHint:aError.errorDescription];
            
        }
    }];
}



#pragma mark - Table view data source
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    if (self.notificationType == AgoraNotificationSettingTypeSelf) {
        return 2;
    }
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    if (self.notificationType == AgoraNotificationSettingTypeSelf) {
        if (section == 0) {
            return 3;
        }
        return 2;
        
    }else if (self.notificationType == AgoraNotificationSettingTypeSingleChat)
    {
        return 2;
    }else{
        return 2;
    }
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
    if (self.notificationType == AgoraNotificationSettingTypeSelf) {
        return 30.0f;
    }
    return 0;
}

- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section {
    if (self.notificationType != AgoraNotificationSettingTypeSelf) {
        return [UIView new];
    }
    UIView *sectionView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, KScreenWidth, 30.0f)];
    
    UILabel *label = [self sectionTitleLabel];
    if (section == 0) {
        label.text = @"Push Notifications";
    }else {
        label.text = @"In-App Notifications";
    }
    [sectionView addSubview:label];
    [label mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.equalTo(sectionView);
        make.left.equalTo(sectionView).offset(16.0);
    }];
    
    return sectionView;
}

- (UILabel *)sectionTitleLabel {
    UILabel *label = [[UILabel alloc] init];
    label.font = [UIFont systemFontOfSize:12.0f];
    label.textColor = TextLabelGrayColor;
    label.textAlignment = NSTextAlignmentLeft;
    return label;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell * cell;
    if (indexPath.section == 0) {
        cell = self.cells[indexPath.row];
    }else{
        cell =self.cells[indexPath.row + 3];
    }
    
    if (cell == _muteCell) {
        [self changeMuteCell];
    }
    if (cell == _remindTypeCell) {
        [self changeRemindCell];
    }
    if (cell == _showPreTextCell) {
        
        [_showPreTextCell.aSwitch setOn:[AgoraChatClient sharedClient].pushManager.pushOptions.displayStyle == AgoraChatPushDisplayStyleMessageSummary];
    }
    if (cell == _vibrateCell) {
        [_vibrateCell.aSwitch setOn:[ACDDemoOptions sharedOptions].playVibration];
    }
    if (cell == _soundCell) {
        [_soundCell.aSwitch setOn:[ACDDemoOptions sharedOptions].playNewMsgSound];
    }
    return cell;
}


#pragma mark Change cell data
- (void)changeMuteCell
{
    NSTimeInterval timestampNow = [[NSDate date] timeIntervalSince1970] * 1000;
    if (self.silentModeItem.expireTimestamp == 0 || self.silentModeItem.expireTimestamp < timestampNow) {
        if(self.notificationType == AgoraNotificationSettingTypeSelf)
        {
            self.muteCell.detailLabel.text = @"Turn On";
        }else{
            self.muteCell.detailLabel.text = @"Mute";
        }
        self.muteCell.subDetailLabel.text =  @"";
        self.muteCell.showSubDetailLabel = NO;
    }else{
        if(self.notificationType == AgoraNotificationSettingTypeSelf)
        {
            self.muteCell.detailLabel.text = @"Turn Off";
        }else{
            self.muteCell.detailLabel.text = @"Unmute";
        }
        self.muteCell.subDetailLabel.text = [ACDDateHelper stringMonthEnglishFromTimestamp:self.silentModeItem.expireTimestamp];
        self.muteCell.showSubDetailLabel = YES;
    }
}

- (void)changeRemindCell
{
    
    NSString * typeStr = @"All Messages";
    switch (self.silentModeItem.remindType) {
        case AgoraChatPushRemindTypeAll:
            typeStr = @"All Messages";
            break;
        case AgoraChatPushRemindTypeMentionOnly:
            typeStr = @"Only @Metions";
            break;
        case AgoraChatPushRemindTypeNone:
            typeStr = @"Nothing";
            break;
        default:
            break;
    }
    
    if (self.notificationType != AgoraNotificationSettingTypeSelf) {
        if (!self.silentModeItem.isConversationRemindTypeEnabled) {
            typeStr = @"Default";
        }
    }
    self.remindTypeCell.detailLabel.text = typeStr;
}


#pragma mark getter and setter

- (UITableView *)table {
    if (_table == nil) {
        _table     = [[UITableView alloc] initWithFrame:CGRectMake(0, 0, KScreenWidth, KScreenHeight)];
        _table.delegate        = self;
        _table.dataSource      = self;
        _table.separatorStyle  = UITableViewCellSeparatorStyleNone;
        _table.keyboardDismissMode = UIScrollViewKeyboardDismissModeOnDrag;
        _table.backgroundColor = COLOR_HEX(0xFFFFFF);
        _table.rowHeight = 54.0f;
    }
    return _table;
}

-(ACDTitleDetailCell *)remindTypeCell
{
    if (_remindTypeCell == nil) {
        _remindTypeCell = [[ACDTitleDetailCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:[ACDTitleDetailCell reuseIdentifier]];
        _remindTypeCell.nameLabel.text = self.remindCellNameTitle;
        _remindTypeCell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
        ACD_WS
        _remindTypeCell.tapCellBlock = ^{
            [weakSelf remindTypeAction];
        };
    }
    return _remindTypeCell;
}

-(ACDSubDetailCell *)muteCell{
    if (_muteCell == nil) {
        _muteCell = [[ACDSubDetailCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:[ACDSubDetailCell reuseIdentifier]];
        _muteCell.nameLabel.text = self.muteCellNameTitle;
        ACD_WS
        _muteCell.tapCellBlock = ^{
            [weakSelf muteAction];
        };
    }
    return _muteCell;
}
-(ACDNameSwitchCell *)showPreTextCell{
    if (_showPreTextCell == nil) {
        _showPreTextCell = [[ACDNameSwitchCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:[ACDNameSwitchCell reuseIdentifier]];
        _showPreTextCell.nameLabel.text = @"Show Preview Text";
        ACD_WS
        _showPreTextCell.switchActionBlock = ^(BOOL isOn) {
            [weakSelf showPreTextAction];
        };
    }
    return _showPreTextCell;
}
-(ACDNameSwitchCell *)soundCell{
    if (_soundCell == nil) {
        _soundCell = [[ACDNameSwitchCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:[ACDNameSwitchCell reuseIdentifier]];
        _soundCell.nameLabel.text = @"Alert Sound";
        _soundCell.switchActionBlock = ^(BOOL isOn) {
            [ACDDemoOptions sharedOptions].playNewMsgSound = isOn;
        };
    }
    return _soundCell;
}
-(ACDNameSwitchCell *)vibrateCell{
    if (_vibrateCell == nil) {
        _vibrateCell = [[ACDNameSwitchCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:[ACDNameSwitchCell reuseIdentifier]];
        _vibrateCell.nameLabel.text = @"Vibrate";
        _vibrateCell.switchActionBlock = ^(BOOL isOn) {
            [ACDDemoOptions sharedOptions].playVibration = isOn;
        };
    }
    return _vibrateCell;
}
-(NSArray *)cells
{
    if (_cells == nil) {
        switch (self.notificationType) {
            case AgoraNotificationSettingTypeSelf:
                _cells = @[self.muteCell,self.remindTypeCell,self.showPreTextCell,self.soundCell,self.vibrateCell];
                break;
            case AgoraNotificationSettingTypeSingleChat:
                _cells = @[self.muteCell,self.remindTypeCell];
                break;
            case AgoraNotificationSettingTypeGroup:
                _cells = @[self.muteCell,self.remindTypeCell];
                break;
            case AgoraNotificationSettingTypeThread:
                _cells = @[self.muteCell,self.remindTypeCell];
                break;
            default:
                break;
        }
    }
    return _cells;
}

-(NSString *)navTitle
{
    if (_navTitle == nil) {
        switch (self.notificationType) {
            case AgoraNotificationSettingTypeSelf:
                _navTitle = @"Notifications";
                self.muteCellNameTitle = @"Do Not Disturb";
                self.remindCellNameTitle = @"Notification Setting";
                break;
            case AgoraNotificationSettingTypeSingleChat:
                _navTitle = @"Contact Notifications";
                self.muteCellNameTitle = @"Mute this Contact";
                self.remindCellNameTitle = @"Notification Setting";
                break;
            case AgoraNotificationSettingTypeGroup:
                _navTitle = @"Group Notifications";
                self.muteCellNameTitle = @"Mute this Group";
                self.remindCellNameTitle = @"Frequency";
                break;
            case AgoraNotificationSettingTypeThread:
                _navTitle = @"Thead Notifications";
                self.muteCellNameTitle = @"Mute this Thread";
                self.remindCellNameTitle = @"Frequency";
                break;
            default:
                break;
        }
    }
    return _navTitle;
}
/*
 #pragma mark - Navigation
 
 // In a storyboard-based application, you will often want to do a little preparation before navigation
 - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
 // Get the new view controller using [segue destinationViewController].
 // Pass the selected object to the new view controller.
 }
 */

@end

File: AgoraChat-Swift/AgoraChat-Swift/code/Controllers/Settings/NotificationsSetting/ACDNotificationSettingViewController.swift
//
//  ACDNotificationSettingViewController.swift
//  AgoraChat-Swift
//
//  Created by 冯钊 on 2022/10/18.
//

import UIKit
import SnapKit

@objc enum AgoraNotificationSettingType: Int32 {
    case me
    case singleChat
    case group
    case thread
}

@objcMembers
class ACDNotificationSettingViewController: UITableViewController {
 
    private var silentModeItem: AgoraChatSilentModeResult?
    private let muteCell: ACDSubDetailCell = Bundle.main.loadNibNamed("ACDSubDetailCell", owner: nil)?.first as! ACDSubDetailCell
    private let remindTypeCell: ACDSubDetailCell = Bundle.main.loadNibNamed("ACDSubDetailCell", owner: nil)?.first as! ACDSubDetailCell
    private lazy var showPreTextCell: ACDNameSwitchCell = {
        let cell = ACDNameSwitchCell(style: .default, reuseIdentifier: nil)
        cell.title = "Show Preview Text".localized
        cell.switchActionHandle = { [unowned self] isOn in
            self.showPreTextAction()
        }
        return cell
    }()
    private lazy var soundCell: ACDNameSwitchCell = {
        let cell = ACDNameSwitchCell(style: .default, reuseIdentifier: nil)
        cell.title = "Alert Sound".localized
        cell.switchActionHandle = { isOn in
            ACDDemoOptions.shared.playNewMsgSound = isOn
        }
        return cell
    }()
    private lazy var vibrateCell: ACDNameSwitchCell = {
        let cell = ACDNameSwitchCell(style: .default, reuseIdentifier: nil)
        cell.title = "Vibrate".localized
        cell.switchActionHandle = { isOn in
            ACDDemoOptions.shared.playVibration = isOn
        }
        return cell
    }()
    
    var conversationID: String?
    var notificationType: AgoraNotificationSettingType = .me
    private var dateFormatter = DateFormatter()
    
    override func viewDidLoad() {
        super.viewDidLoad()

        self.view.backgroundColor = .white
        
        var title: String
        var muteCellNameTitle: String
        var remindCellNameTitle: String
        switch self.notificationType {
        case .singleChat:
            title = "Contact Notifications".localized
            muteCellNameTitle = "Mute this Contact".localized
            remindCellNameTitle = "Notification Setting".localized
        case .group:
            title = "Group Notifications".localized
            muteCellNameTitle = "Mute this Group".localized
            remindCellNameTitle = "Frequency".localized
        case .thread:
            title = "Thread Notifications".localized
            muteCellNameTitle = "Mute this Thread".localized
            remindCellNameTitle = "Frequency".localized
        default:
            title = "Notifications".localized
            muteCellNameTitle = "Do Not Disturb".localized
            remindCellNameTitle = "Notification Setting".localized
        }
        
        self.navigationItem.leftBarButtonItem = ACDUtil.customLeftButtonItem(title: title, action: #selector(backAction), target: self)
        
        self.tableView.separatorStyle = .none
        self.tableView.keyboardDismissMode = .onDrag
        self.tableView.backgroundColor = .white
        self.tableView.rowHeight = 54.0
        if #available(iOS 15.0, *) {
            self.tableView.sectionHeaderTopPadding = 0.0
        }
        
        self.muteCell.nameLabel.text = muteCellNameTitle
        self.muteCell.accessoryType = .disclosureIndicator
        self.muteCell.tapCellHandle = { [unowned self] in
            self.muteAction()
        }
        
        self.remindTypeCell.nameLabel.text = remindCellNameTitle
        self.remindTypeCell.accessoryType = .disclosureIndicator
        self.remindTypeCell.tapCellHandle = { [unowned self] in
            self.remindTypeAction()
        }
        
        self.dateFormatter.dateFormat = "MMM dd,yyyy, HH:mm".localized
        
        self.getDataFromSever()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.navigationController?.setNavigationBarHidden(false, animated: true)
    }
    
    private func getDataFromSever() {
        MBProgressHUD.showAdded(to: self.view, animated: true)
        if self.notificationType == AgoraNotificationSettingType.me {
            AgoraChatClient.shared().pushManager?.getSilentModeForAll(completion: { result, error in
                MBProgressHUD.hide(for: self.view, animated: true)
                if let error = error {
                    self.showHint(error.errorDescription)
                } else {
                    self.silentModeItem = result
                    self.tableView.reloadData()
                }
            })
        } else if let conversationID = self.conversationID {
            let chatType: AgoraChatConversationType = self.notificationType == .singleChat ? .chat : .groupChat
            AgoraChatClient.shared().pushManager?.getSilentMode(forConversation: conversationID, conversationType: chatType, completion: { result, error in
                MBProgressHUD.hide(for: self.view, animated: true)
                if let error = error {
                    self.showHint(error.errorDescription)
                } else {
                    self.silentModeItem = result
                    self.tableView.reloadData()
                }
            })
        }
    }
    
    @objc private func backAction() {
        self.navigationController?.popViewController(animated: true)
    }
    
    private func remindTypeDefault() {
        guard let conversationID = self.conversationID else {
            return
        }
        MBProgressHUD.showAdded(to: self.view, animated: true)
        let type: AgoraChatConversationType = self.notificationType == .singleChat ? .chat : .groupChat
        AgoraChatClient.shared().pushManager?.clearRemindType(forConversation: conversationID, conversationType: type, completion: { result, error in
            MBProgressHUD.hide(for: self.view, animated: true)
            if let error = error {
                self.showHint(error.errorDescription)
            } else {
                self.silentModeItem = result
                self.tableView.reloadData()
            }
        })
    }
    
    private func remindTypeChange(type: AgoraChatPushRemindType) {
        let param = AgoraChatSilentModeParam(paramType: .remindType)
        param.remindType = type
        self.silentModeChange(param: param)
    }
    
    private func silentModeChange(param: AgoraChatSilentModeParam) {
        if self.notificationType == .me {
            AgoraChatClient.shared().pushManager?.setSilentModeForAll(param, completion: { result, error in
                if let error = error {
                    self.showHint(error.errorDescription)
                } else {
                    self.silentModeItem = result
                    self.tableView.reloadData()
                }
            })
        } else if let conversationID = self.conversationID {
            let type: AgoraChatConversationType = self.notificationType == .singleChat ? .chat : .groupChat
            AgoraChatClient.shared().pushManager?.setSilentModeForConversation(conversationID, conversationType: type, params: param, completion: { result, error in
                if let error = error {
                    self.showHint(error.errorDescription)
                } else {
                    self.silentModeItem = result
                    self.tableView.reloadData()
                }
            })
        }
    }
    
    private func remindTypeAction() {
        let vc = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
        if self.notificationType != .me {
            let action = UIAlertAction(title: "Default".localized, style: .default) { _ in
                self.remindTypeDefault()
            }
            action.setValue(UIColor.black, forKey: "titleTextColor")
            vc.addAction(action)
        }
        let allAction = UIAlertAction(title: "All Messages".localized, style: .default) { _ in
            self.remindTypeChange(type: .all)
        }
        allAction.setValue(UIColor.black, forKey: "titleTextColor")
        vc.addAction(allAction)
        
        if self.notificationType != .singleChat {
            let action = UIAlertAction(title: "Only @Metions".localized, style: .default) { _ in
                self.remindTypeChange(type: .mentionOnly)
            }
            action.setValue(UIColor.black, forKey: "titleTextColor")
            vc.addAction(action)
        }
        
        let noneAction = UIAlertAction(title: "Nothing".localized, style: .default) { _ in
            self.remindTypeChange(type: .none)
        }
        noneAction.setValue(UIColor.black, forKey: "titleTextColor")
        vc.addAction(noneAction)
            
        vc.addAction(UIAlertAction(title: LocalizedString.Cancel, style: .cancel))
        
        self.present(vc, animated: true)
    }

    private func muteAction() {
        if let expireTimestamp = self.silentModeItem?.expireTimestamp, expireTimestamp > Date().timeIntervalSince1970 * 1000 {
            let mute = self.muteCell.subDetailLabel.text ?? ""
            var titleStr = "Turn off Do Not Disturb?".localized
            var infoStr = String(format: "You have set Do Not Disturb until \n %@".localized, mute)
            if self.notificationType != .me {
                if self.notificationType == .group {
                    titleStr = "Mute this Group".localized
                    infoStr = String(format: "You have muted this Group until \n %@".localized, mute)

                } else {
                    titleStr = "Mute this Thread".localized
                    infoStr = String(format: "You have muted this Thread until \n %@".localized, mute)
                }
            }
            let vc = UIAlertController(title: titleStr, message: infoStr, preferredStyle: .alert)
            vc.addAction(UIAlertAction(title: LocalizedString.Cancel, style: .cancel))
            vc.addAction(UIAlertAction(title: LocalizedString.Ok, style: .default) { _ in
                self.closeMute()
            })
            self.present(vc, animated: true)
        } else {
            let vc = ACDSilentModeSetViewController(conversationID: self.conversationID, notificationType: self.notificationType) { item in
                self.silentModeItem = item
                self.tableView.reloadData()
            }
            self.navigationController?.pushViewController(vc, animated: true)
        }
    }
    
    private func closeMute() {
        let param = AgoraChatSilentModeParam(paramType: .duration)
        param.silentModeDuration = 0
        self.silentModeChange(param: param)
    }

    private func showPreTextAction() {
        let style: AgoraChatPushDisplayStyle = self.showPreTextCell.switch.isOn ? .messageSummary : .simpleBanner
        AgoraChatClient.shared().pushManager?.update(style) { error in
//            let error: AgoraChatError? = error
//            if let error = error {
//                self.showHint(error.errorDescription)
//            }
        }
    }

    override func numberOfSections(in tableView: UITableView) -> Int {
        return self.notificationType == .me ? 2 : 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if (self.notificationType == .me) {
            if section == 0 {
                return 2
            }
            return 2
        } else if (self.notificationType == .singleChat) {
            return 2
        } else {
            return 2
        }
    }
    
    override func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return self.notificationType == .me ? 30 : 0
    }
    
    override func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        if (self.notificationType != .me) {
            return UIView()
        }
        let view = UIView(frame: CGRect(x: 0, y: 0, width: UIScreen.main.bounds.width, height: 30))
        let label = UILabel()
        label.font = UIFont.systemFont(ofSize: 12)
        label.textColor = .gray
        label.textAlignment = .left
        label.text = section == 0 ? "Push Notifications".localized : "In-App Notifications".localized
        
        view.addSubview(label)
        label.snp.makeConstraints { make in
            make.centerY.equalTo(view)
            make.left.equalTo(16)
        }
        return view
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if indexPath.section == 0 {
            if indexPath.row == 0 {
                self.changeMuteCell()
                return self.muteCell
            } else if indexPath.row == 1 {
                self.changeRemindCell()
                return self.remindTypeCell
            } else {
                self.showPreTextCell.isOn = AgoraChatClient.shared().pushManager?.pushOptions?.displayStyle == .messageSummary
                return self.showPreTextCell
            }
        } else {
            if indexPath.row == 0 {
                self.soundCell.isOn = ACDDemoOptions.shared.playNewMsgSound
                return self.soundCell
            } else {
                self.vibrateCell.isOn = ACDDemoOptions.shared.playVibration
                return self.vibrateCell
            }
        }
    }
    
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        if indexPath.section == 0 && indexPath.row == 0 {
            return self.muteCell.height
        }
        return 54
    }
    
    private func changeMuteCell() {
        if let expireTimestamp = self.silentModeItem?.expireTimestamp, expireTimestamp > Date().timeIntervalSince1970 * 1000 {
            self.muteCell.detailLabel.text = self.notificationType == .me ? "Turn Off".localized : "Unmute".localized
            
            let date = Date(timeIntervalSince1970: expireTimestamp / 1000)
            self.muteCell.subDetailLabel.text = self.dateFormatter.string(from: date)
            self.muteCell.showSubDetailLabel = true
        } else {
            self.muteCell.detailLabel.text = self.notificationType == .me ? "Turn On".localized : "Mute".localized
            self.muteCell.subDetailLabel.text = ""
            self.muteCell.showSubDetailLabel = false
        }
    }
    
    private func changeRemindCell() {
        var typeStr: String
        switch self.silentModeItem?.remindType {
        case .all:
            typeStr = "All Messages".localized
        case .mentionOnly:
            typeStr = "Only @Metions".localized
        default:
            typeStr = "Nothing".localized
        }
        if self.notificationType != .me {
            if self.silentModeItem?.isConversationRemindTypeEnabled != true {
                typeStr = "Default".localized
            }
        }
        self.remindTypeCell.detailLabel.text = typeStr
    }
}
